### ISSUE TITLE
Confusions with dynamic paired collection creation in tools - tool wrapper XML

### ISSUE THREAD

**Issue by blankclemens** [galaxyproject/galaxy#4573]
While trying to produce dynamic paired lists from a tool I encountered some issues and came across an interesting behavior where I'm not sure if it's due to my limited understanding of how the underlying things work, or if it is indeed weird:

I tried to follow and apply [this example file](https://github.com/galaxyproject/galaxy/blob/master/test/functional/tools/collection_creates_dynamic_list_of_pairs.xml) to my wrapper, and for some reason it never wanted to work out. It seems that having a paired collection as an input somehow changes how the tool works - adding a dummy paired collection input to the above mentioned example file changes the output, even though in my expectation nothing really should have changed. You can try it if you add for example the following input:

```
<param name="dummy_collection_input" format="fastq" type="data_collection" collection_type="paired" label="FASTQ Paired Dataset"/> 
```

So I'm wondering: What is this dummy input changing in this example file that it won't create the `Duplicate list` anymore (or as expected multiple times for each item in the input collection), but instead only the hidden single elements?

@jmchilton Maybe you can shed some light here as the author of #538

**Comment by nsoranzo** [comment]
```
format="fastq,gz"
```
should be
```
format="fastq.gz"
```

Not sure that's the cause of your issue.

**Comment by blankclemens** [comment]
Thanks, my bad.

The issue persists though.

**Comment by blankclemens** [comment]
Status update:

It seems like having a collection as input, regardless if it is a dummy like in this case, the output collection will become `"collection_type": "list:list:paired"`, which is probably not a valid type and will not be shown in the graphical history (I thought it doesn't exist until I looked in the API).

Does it automatically add `list:` to the output collection type? 

Edit: Seems like it. If I change the collection output just to `paired` it will result in a (not working) `list:paired` in the history, being yellow and not turning green, cause of `AssertionError: Cannot add nested collection to collection of type [CollectionTypeDescription[paired]]`.

So having as input a `collection_type="paired"` adds a `list:` to the output collection which will not work if the output is already a `list:paired`? @jmchilton

**Comment by jmchilton** [comment]
@blankclemens So you started with https://github.com/galaxyproject/galaxy/blob/master/test/functional/tools/collection_creates_dynamic_list_of_pairs.xml and you added a paired input param? I think what happened made sense - Galaxy (and me) just isn't good at explaining it.

Lets say you have a list of paired datasets - so lets say 3 forward-reverse pairs - and you send that to a tool consumes a ``paired`` collection_type input . Well then the tool will consume the pairs - but the tool needs to be run three times (since the tool consumes pairs not lists of pairs) - and so those results are going to be strung together into a list with the same element identifiers and indices as that initial paired list. Since each output of this tool is a list of pairs - the result will be a list of list of pairs or ``list:list:paired`` - a completely valid and sometimes useful datatype.

Lets say you have some data in pairs of BAM files where each BAM file contains a bunch of technical replicates for the same biological sample and you have a list of samples. So you start with ``list:paired`` - where the outer list is indexed by "sample". Now lets say you have a tool that looks at a pair of bam files and splits it up based on technical replicate - this would be a tool like we described above that breaks ``paired``s dynamicall into ``list:paired``s. If I take a list of sample pairs and run it over a tool that takes each pair and produces a list of bam files split on technical replicates - I would expect the result to be a list of samples where each list contains a list of dataset pairs for each replicate (a ``list:list:paired``).

This is very likely not the tool you are trying to write however - ``paired -> list:paired`` is a valid though not common transformation. I'd expect what you want is either ``list:paired -> list:paired`` or ``dataset -> dataset``. Can you answer a couple quick questions - they'd help figure this out I think:

- Can you describe the transformation you are hoping to accomplish at sort of a high-level and then what the actual tool expects for inputs and outputs?
- Does the underlying tool really need to process dataset in pairs - or can it look at the files individually and let Galaxy string the results together?
- If I have a list of paired inputs that I want to send to this tool, can you determine the number of elements that will be produced ahead of time or is it truly dynamic?

**Comment by blankclemens** [comment]
Thanks @jmchilton for this long explanation. While `list:list:paired` sounds useful to me, I thought it's not valid since it wasn't displayed in the history - I just found it looking at the history element through the API.

Back to my specific problem and your questions:

- FASTA files should be split up to multiple files according to a certain Barcode list. Therefore the input can be single and paired end elements or collections. For each paired input the output should be a paired collection. Having two pairs in a paired collection should therefore lead to two collections with the split up pairs (still in pairs). As of know, I might end up with hundreds of uncollected output files which is rather uncomfortable to continue with.

- In my understanding yes, since it needs to perform the same actions on the corresponding mates.

- Theoretically it would be possible to calculate the amount of output files according to the list of Barcodes, but it not only can be quite a high number, it also might vary according to some options - but it is deterministic, yes.

While changing the original tool wrapper I came across some problems which made me aware I'm having some issues understanding how it works, that's why I though I best start from the beginning with your example files.

**Comment by jmchilton** [comment]
Yeah there are issues with ``list:list:paired`` collections in the GUI - but they are valid in some ways. So it seems like this tool really does map ``list:paired`` to ``list:list:paired`` - maybe you just need to wait for Galaxy to catch up in terms of representing these collections in the GUI. If you can map ``list:paired`` to ``list:paired`` maybe combining all the barcodes for a particular pair of files into the same list - Galaxy can do that by setting the output collection type to ``list:paired`` and specifying an input of type ``type="data_collection" collection_type="paired,list:paired"``. You'd maybe want two variant of your tool then - one that consumes and produces paired data and one that consumes and produces single strand data.
